File: 1597_build_binary_expression_tree_from_infix_expression/solution.go
// LeetCode 1597 - Build Binary Expression Tree From Infix Expression
// https://leetcode.com/problems/build-binary-expression-tree-from-infix-expression/

type Node struct {
	Val   byte
	Left  *Node
	Right *Node
}

func expTree(s string) *Node {
	nodes := []*Node{}
	ops := []byte{}
	priority := map[byte]int{'+': 1, '-': 1, '*': 2, '/': 2}
	apply := func() {
		op := ops[len(ops)-1]
		ops = ops[:len(ops)-1]
		right := nodes[len(nodes)-1]
		nodes = nodes[:len(nodes)-1]
		left := nodes[len(nodes)-1]
		nodes = nodes[:len(nodes)-1]
		nodes = append(nodes, &Node{Val: op, Left: left, Right: right})
	}
	for i := 0; i < len(s); i++ {
		ch := s[i]
		if ch >= '0' && ch <= '9' {
			nodes = append(nodes, &Node{Val: ch})
		} else if ch == '(' {
			ops = append(ops, ch)
		} else if ch == ')' {
			for ops[len(ops)-1] != '(' {
				apply()
			}
			ops = ops[:len(ops)-1]
		} else {
			for len(ops) > 0 && ops[len(ops)-1] != '(' && priority[ops[len(ops)-1]] >= priority[ch] {
				apply()
			}
			ops = append(ops, ch)
		}
	}
	for len(ops) > 0 {
		apply()
	}
	return nodes[0]
}

File: 1597_build_binary_expression_tree_from_infix_expression/Solution.kt
// LeetCode 1597 - Build Binary Expression Tree From Infix Expression
// https://leetcode.com/problems/build-binary-expression-tree-from-infix-expression/

import java.util.ArrayDeque

class Node {
    var `val`: Char = ' '
    var left: Node? = null
    var right: Node? = null

    constructor(`val`: Char) {
        this.`val` = `val`
    }

    constructor(`val`: Char, left: Node?, right: Node?) {
        this.`val` = `val`
        this.left = left
        this.right = right
    }
}

class Solution {
    fun expTree(s: String): Node? {
        val nodes = ArrayDeque<Node>()
        val ops = ArrayDeque<Char>()
        val priority = mapOf('+' to 1, '-' to 1, '*' to 2, '/' to 2)
        for (ch in s) {
            when {
                ch in '0'..'9' -> nodes.push(Node(ch))
                ch == '(' -> ops.push(ch)
                ch == ')' -> {
                    while (ops.peek() != '(') apply(nodes, ops)
                    ops.pop()
                }
                else -> {
                    while (ops.isNotEmpty() && ops.peek() != '(' &&
                        priority[ops.peek()]!! >= priority[ch]!!
                    ) {
                        apply(nodes, ops)
                    }
                    ops.push(ch)
                }
            }
        }
        while (ops.isNotEmpty()) apply(nodes, ops)
        return nodes.peek()
    }

    private fun apply(nodes: ArrayDeque<Node>, ops: ArrayDeque<Char>) {
        val op = ops.pop()
        val right = nodes.pop()
        val left = nodes.pop()
        nodes.push(Node(op, left, right))
    }
}
